### PR TITLE
 add context namespacing filter (Bug 1196022)

### PR DESCRIPTION
**Pull request by deads2k**
namespaces not set in context.  Do not merge this without additional integration tests to avoid this problem in the future.  

Integration tests will be much easier with https://github.com/openshift/origin/pull/1128.  I'm going to try that route first.

### CODE REVIEW

**Comment by deads2k** [comment]
@liggitt review?  It came out a little bigger than I wanted, but I think it's still reviewable.

File: test/integration/authorization_test.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	openshiftClient, openshiftClientConfig, err := startConfig.GetOpenshiftClient()
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	removeInsecureOptions := &policy.RemoveGroupOptions{
```
Comment (liggitt) [inline test/integration/authorization_test.go]: add a TODO to remove once policy gets tightened

Comment (deads2k) [inline test/integration/authorization_test.go]: done

File: test/integration/test_server.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+	go func() {
+		err := startConfig.Start(args)
+		if err != nil {
+			fmt.Printf("ERROR STARTING SERVER! %v", err)
+		}
+	}()
+
+	// wait for the server to come up
+	if err := cmdutil.WaitForSuccessfulDial(true, "tcp", masterAddr, 100*time.Millisecond, 100*time.Millisecond, 100); err != nil {
```
Comment (liggitt) [inline test/integration/test_server.go]: is 20 seconds long enough to wait?

Comment (deads2k) [inline test/integration/test_server.go]: > is 20 seconds long enough to wait?

Takes 4 on my machine.

File: test/integration/test_server.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+		return *startConfig, err
+	}
+
+	stopChannel := make(chan struct{})
```
Comment (liggitt) [inline test/integration/test_server.go]: we definitely want a timeout here... an integration test should just fail within a reasonable amount of time, not hang

Comment (deads2k) [inline test/integration/test_server.go]: I thought we had a global timeout for tests of 45 seconds.  Do we not?

File: test/integration/test_server.go [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+	startConfig.MasterAddr.Set(masterAddr)
+	startConfig.BindAddr.Set(masterAddr)
+	startConfig.EtcdAddr.Set(getEtcdURL())
+
+	startConfig.Complete(args)
+
+	go func() {
+		err := startConfig.Start(args)
+		if err != nil {
+			fmt.Printf("ERROR STARTING SERVER! %v", err)
```
Comment (liggitt) [inline test/integration/test_server.go]: seems like we should do something more drastic here to fail fast

File: pkg/cmd/server/origin/master.go [hunk 5]
```diff
@@ -692,3 +693,26 @@ type clientDeploymentInterface struct {
...
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx, ok := contextMapper.Get(req)
+		if !ok {
+			http.Error(w, "Unable to find request context", http.StatusInternalServerError)
+			return
+		}
+
+		if _, exists := kapi.NamespaceFrom(ctx); !exists {
+			requestInfo, _ := infoResolver.GetAPIRequestInfo(req)
+			ctx = kapi.WithNamespace(ctx, requestInfo.Namespace)
```
Comment (liggitt) [inline pkg/cmd/server/origin/master.go]: feels like we shouldn't set namespace on the context if GetAPIRequestInfo returned an error or if requestInfo.Namespace is ""

File: test/integration/test_server.go (hunk 2 above)
Comment (deads2k) [inline test/integration/test_server.go]: Updated

File: test/integration/test_server.go (hunk 4 above)
Comment (deads2k) [inline test/integration/test_server.go]: updated.

File: pkg/cmd/server/origin/master.go (hunk 5 above)
Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: An error doesn't necessarily mean that a namespace wasn't found.  From the godoc:
```
If error is not nil, APIRequestInfo holds the information as best it is known before the failure
```

An empty namespace is not the same as a missing namespace.  Empty namespace can mean any namespace, all namespaces, or no namespace depending upon context.  It is up to the namespace user to decide which is appropriate.

File: test/integration/authorization_test.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+
+	_, err = haroldClient.Deployments("hammer-project").List(labels.Everything(), labels.Everything())
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	_, err = markClient.Deployments("hammer-project").List(labels.Everything(), labels.Everything())
+	if (err == nil) || (!strings.Contains(err.Error(), "Forbidden")) {
```
Comment (liggitt) [inline test/integration/authorization_test.go]: can you use `kerrors.IsForbidden(err)`? I'm never sure what type of error we're getting back at each layer

File: pkg/cmd/server/origin/master.go [hunk 7]
```diff
@@ -692,3 +693,34 @@ type clientDeploymentInterface struct {
...
+		}
+
+		if _, exists := kapi.NamespaceFrom(ctx); !exists {
+			requestInfo, _ := infoResolver.GetAPIRequestInfo(req)
+
+			// TODO reconsider special casing this.  Having the special case hereallow us to fully share the kube
+			// APIRequestInfoResolver without any modification or customization.
+			namespace := requestInfo.Namespace
+			if (requestInfo.Resource == "projects") && (len(requestInfo.Name) > 0) {
+				namespace = requestInfo.Name
```
Comment (liggitt) [inline pkg/cmd/server/origin/master.go]: if requestInfo has a namespace, and we get in here, is that an error?

Comment (deads2k) [inline pkg/cmd/server/origin/master.go]: Errors aren't caught on other weird conditions with non-namespaced resources having a namespace, so I'm going to with no.

Comment (liggitt) [inline pkg/cmd/server/origin/master.go]: yeah, that's concerning, actually... means you can do GET /namespaces/foo/oauthaccesstokens, get through authz if you have permission within namespace `foo`, then the rest layer ignores the namespace. not a problem to fix here, though

File: pkg/cmd/server/origin/master.go [hunk 8]
```diff
@@ -692,3 +693,34 @@ type clientDeploymentInterface struct {
...
+
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		ctx, ok := contextMapper.Get(req)
+		if !ok {
+			http.Error(w, "Unable to find request context", http.StatusInternalServerError)
+			return
+		}
+
+		if _, exists := kapi.NamespaceFrom(ctx); !exists {
+			requestInfo, _ := infoResolver.GetAPIRequestInfo(req)
```
Comment (liggitt) [inline pkg/cmd/server/origin/master.go]: still feels weird to set an empty namespace on the context regardless of the success of GetAPIRequestInfo.

If GetAPIRequestInfo() fails, and requestInfo.Namespace is empty, what does setting an empty namespace communicate to downstream consumers of the context?

File: test/integration/authorization_test.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+		t.Errorf("unexpected error: %v", err)
+	}
+	// TODO make kube and origin authorization failures cause a kapierror.Forbidden
+	_, err = markClient.Projects().Get("hammer-project")
+	if (err == nil) || (!strings.Contains(err.Error(), "Forbidden")) {
+		t.Errorf("expected forbidden error, but didn't get one")
+	}
+
+	// wait for the project authorization cache to catch the change.  It is on a one second period
+	time.Sleep(1 * time.Second)
```
Comment (liggitt) [inline test/integration/authorization_test.go]: is sleeping the exact sync period going to be fragile?

File: test/integration/authorization_test.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+	haroldProjects, err := haroldClient.Projects().List(labels.Everything(), labels.Everything())
+	if err != nil {
+		t.Errorf("unexpected error: %v", err)
+	}
+	if !((len(haroldProjects.Items) == 1) && (haroldProjects.Items[0].Name == "hammer-project")) {
```
Comment (liggitt) [inline test/integration/authorization_test.go]: I would find this easier to read, but maybe it's just me
`if (len(haroldProjects.Items) != 1) || (haroldProjects.Items[0].Name != "hammer-project") {`

File: test/integration/test_server.go [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+func init() {
+	requireEtcd()
+}
+
+func StartTestServer(args ...string) (start.Config, error) {
+	deleteAllEtcdKeys()
+
+	startConfig := start.NewDefaultConfig()
+
+	basedir := path.Join(os.TempDir(), "openshift-integration-tests")
```
Comment (liggitt) [inline test/integration/test_server.go]: is it possible to put this somewhere that `make clean` will clean up?

File: test/integration/test_server.go [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	adminClientConfig := clientConfig
+	adminClientConfig.BearerToken = token
+	adminClientConfig.TLSClientConfig.CertFile = ""
+	adminClientConfig.TLSClientConfig.KeyFile = ""
+	adminClientConfig.TLSClientConfig.CertData = nil
+	adminClientConfig.TLSClientConfig.KeyData = nil
```
Comment (liggitt) [inline test/integration/test_server.go]: clear Username/Password as well

File: test/integration/authorization_test.go (hunk 10 above)
Comment (deads2k) [inline test/integration/authorization_test.go]: > I would find this easier to read, but maybe it's just me

I tried that before running demorgans on it.  I couldn't read it in that direction.

File: test/integration/test_server.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+}
+
+// StartTestAllInOne starts up a test all-in-one and returns back the startConfig so you can get clients and certs
+func StartTestAllInOne() (start.Config, error) {
+	return StartTestServer()
+}
+
+// CreateNewProject creates a new project using the clusterAdminClient, then gets a token for the adminUser and returns
+// back a client for the admin user
+func CreateNewProject(clusterAdminClient *client.Client, clientConfig kclient.Config, projectName, adminUser string) (*client.Client, error) {
```
Comment (liggitt) [inline test/integration/test_server.go]: would it be worthwhile to return a kclient as well, so we can check kubernetes resources inside the project?

**Comment by liggitt** [comment]
LGTM

**Comment by deads2k** [comment]
[merge]

**Comment by smarterclayton** [comment]
Um... why are we doing this with namespacing filter?  Namespace is set by API calls, not by arbitrary filters.

**Comment by deads2k** [comment]
> Um... why are we doing this with namespacing filter? Namespace is set by API calls, not by arbitrary filters

After removing namespace from authorization.Attributes and instead using the value set on context, we needed to make sure the value was set on context.  Since the API calls don't get invoked until after the call is authorized, we have to retrieve the namespace earlier than that.  Previously, we were simply setting it on a private object, but after we started using the context for namespace values it seemed to make sense to set it in a filter placed before authorization.

**Comment by smarterclayton** [comment]
Can you open an issue upstream to track "check API login in API, and ensure non API calls go through the non-api authorization path"?

**Pull request merged**